In [878]:
#IMPORT LIBRARIES


import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

plt.style.use("seaborn-v0_8")
sns.set_palette("viridis")

# Enable full-width display
pd.set_option("display.max_columns", None)
pd.set_option("display.width", 1500)

In [879]:

# 2. LOAD DATA


# Place your downloaded data in: data/raw/xente_dataset.csv
data_path = r"C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\data\raw\CreditRisk-data.csv"

df = pd.read_csv(data_path)

# View structure
df.head()


TransactionId         BatchId       AccountId       SubscriptionId       CustomerId CurrencyCode  CountryCode    ProviderId     ProductId     ProductCategory    ChannelId   Amount  Value  TransactionStartTime  PricingStrategy  FraudResult
0  TransactionId_76871   BatchId_36123  AccountId_3957   SubscriptionId_887  CustomerId_4406          UGX          256  ProviderId_6  ProductId_10             airtime  ChannelId_3   1000.0   1000  2018-11-15T02:18:49Z                2            0
1  TransactionId_73770   BatchId_15642  AccountId_4841  SubscriptionId_3829  CustomerId_4406          UGX          256  ProviderId_4   ProductId_6  financial_services  ChannelId_2    -20.0     20  2018-11-15T02:19:08Z                2            0
2  TransactionId_26203   BatchId_53941  AccountId_4229   SubscriptionId_222  CustomerId_4683          UGX          256  ProviderId_6   ProductId_1             airtime  ChannelId_3    500.0    500  2018-11-15T02:44:21Z                2            0
3    TransactionId_380  BatchId_102363   AccountId_648  SubscriptionId_2185   CustomerId_988          UGX          256  ProviderId_1  ProductId_21        utility_bill  ChannelId_3  20000.0  21800  2018-11-15T03:32:55Z                2            0
4  TransactionId_28195   BatchId_38780  AccountId_4841  SubscriptionId_3829   CustomerId_988          UGX          256  ProviderId_4   ProductId_6  financial_services  ChannelId_2   -644.0    644  2018-11-15T03:34:21Z                2            0

In [880]:

# 3. BASIC OVERVIEW


print("📌 Dataset Dimensions:")
print(df.shape)

print("\n📌 Column Data Types:")
print(df.dtypes)

print("\n📌 Preview:")
df.head()


📌 Dataset Dimensions:
(95662, 16)

📌 Column Data Types:
TransactionId            object
BatchId                  object
AccountId                object
SubscriptionId           object
CustomerId               object
CurrencyCode             object
CountryCode               int64
ProviderId               object
ProductId                object
ProductCategory          object
ChannelId                object
Amount                  float64
Value                     int64
TransactionStartTime     object
PricingStrategy           int64
FraudResult               int64
dtype: object

📌 Preview:


TransactionId         BatchId       AccountId       SubscriptionId       CustomerId CurrencyCode  CountryCode    ProviderId     ProductId     ProductCategory    ChannelId   Amount  Value  TransactionStartTime  PricingStrategy  FraudResult
0  TransactionId_76871   BatchId_36123  AccountId_3957   SubscriptionId_887  CustomerId_4406          UGX          256  ProviderId_6  ProductId_10             airtime  ChannelId_3   1000.0   1000  2018-11-15T02:18:49Z                2            0
1  TransactionId_73770   BatchId_15642  AccountId_4841  SubscriptionId_3829  CustomerId_4406          UGX          256  ProviderId_4   ProductId_6  financial_services  ChannelId_2    -20.0     20  2018-11-15T02:19:08Z                2            0
2  TransactionId_26203   BatchId_53941  AccountId_4229   SubscriptionId_222  CustomerId_4683          UGX          256  ProviderId_6   ProductId_1             airtime  ChannelId_3    500.0    500  2018-11-15T02:44:21Z                2            0
3    TransactionId_380  BatchId_102363   AccountId_648  SubscriptionId_2185   CustomerId_988          UGX          256  ProviderId_1  ProductId_21        utility_bill  ChannelId_3  20000.0  21800  2018-11-15T03:32:55Z                2            0
4  TransactionId_28195   BatchId_38780  AccountId_4841  SubscriptionId_3829   CustomerId_988          UGX          256  ProviderId_4   ProductId_6  financial_services  ChannelId_2   -644.0    644  2018-11-15T03:34:21Z                2            0

In [881]:
df['TransactionStartTime'].dtype


dtype('O')

In [882]:
import pandas as pd

df['TransactionStartTime'] = pd.to_datetime(df['TransactionStartTime'])


In [883]:
df['TransactionStartTime'].dtype


datetime64[ns, UTC]

In [884]:
df['TransactionStartTime'].isna().sum()

np.int64(0)

In [885]:
df['TransactionStartTime'] = pd.to_datetime(
    df['TransactionStartTime'], 
    errors='coerce'
)


In [886]:
df['transaction_year'] = df['TransactionStartTime'].dt.year
df['transaction_month'] = df['TransactionStartTime'].dt.month
df['transaction_day'] = df['TransactionStartTime'].dt.day
df['transaction_hour'] = df['TransactionStartTime'].dt.hour


In [887]:
# Convert CountryCode to category
df["CountryCode"] = df["CountryCode"].astype("category")

# Verify
df.dtypes


TransactionId                        object
BatchId                              object
AccountId                            object
SubscriptionId                       object
CustomerId                           object
CurrencyCode                         object
CountryCode                        category
ProviderId                           object
ProductId                            object
ProductCategory                      object
ChannelId                            object
Amount                              float64
Value                                 int64
TransactionStartTime    datetime64[ns, UTC]
PricingStrategy                       int64
FraudResult                           int64
transaction_year                      int32
transaction_month                     int32
transaction_day                       int32
transaction_hour                      int32
dtype: object

In [888]:
# Convert CountryCode to categorical type
df['CountryCode'] = df['CountryCode'].astype('category')

# Verify the change
df['CountryCode'].dtype


CategoricalDtype(categories=[256], ordered=False, categories_dtype=int64)

In [889]:
df['CountryCode'].value_counts()


CountryCode
256    95662
Name: count, dtype: int64

In [890]:
import matplotlib.pyplot as plt

# Your plotting code
df['CountryCode'].value_counts().plot(kind='bar', figsize=(8, 4))
plt.title("Distribution of Transactions by Country Code")
plt.xlabel("Country Code")
plt.ylabel("Transaction Count")

# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\transaction_count.png')
plt.close()  # Close the plot to free memory


In [891]:
df['CurrencyCode'].nunique(), df['CurrencyCode'].unique()


(1, array(['UGX'], dtype=object))

In [892]:
df['AccountId'].nunique(), df['CustomerId'].nunique()


(3633, 3742)

In [893]:
df['AccountId'].value_counts().head(10)


AccountId
AccountId_4841    30893
AccountId_4249     4457
AccountId_4840     1738
AccountId_3206     1105
AccountId_318      1070
AccountId_10        965
AccountId_3595      465
AccountId_751       411
AccountId_2648      359
AccountId_2314      356
Name: count, dtype: int64

In [894]:
df['CustomerId'].value_counts().head(10)


CustomerId
CustomerId_7343    4091
CustomerId_3634    2085
CustomerId_647     1869
CustomerId_1096     784
CustomerId_4033     778
CustomerId_3066     612
CustomerId_2528     610
CustomerId_2728     586
CustomerId_3078     573
CustomerId_806      557
Name: count, dtype: int64

- Only one currency (UGX) is present in the dataset.
- This simplifies modeling since no currency normalization is required.
- However, if multiple currencies were expected, this may indicate:
  - Data filtering issues
  - Incomplete ingestion
  - Business scope limited to Uganda


In [895]:
df.describe().T


count         mean            std        min     25%     50%     75%        max
Amount             95662.0  6717.846433  123306.797164 -1000000.0   -50.0  1000.0  2800.0  9880000.0
Value              95662.0  9900.583941  123122.087776        2.0   275.0  1000.0  5000.0  9880000.0
PricingStrategy    95662.0     2.255974       0.732924        0.0     2.0     2.0     2.0        4.0
FraudResult        95662.0     0.002018       0.044872        0.0     0.0     0.0     0.0        1.0
transaction_year   95662.0  2018.497115       0.499994     2018.0  2018.0  2018.0  2019.0     2019.0
transaction_month  95662.0     6.566233       5.224310        1.0     1.0    11.0    12.0       12.0
transaction_day    95662.0    15.902898       8.962822        1.0     8.0    16.0    24.0       31.0
transaction_hour   95662.0    12.447722       4.846964        0.0     8.0    13.0    17.0       23.0

In [896]:
#Raw Distribution
fig, ax = plt.subplots(1, 2, figsize=(14, 5))

sns.histplot(df["Amount"], bins=100, ax=ax[0])
ax[0].set_title("Transaction Amount Distribution")

sns.histplot(df["Value"], bins=100, ax=ax[1])
ax[1].set_title("Transaction Value Distribution")

# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\transaction_value_distribution.png')
plt.close()


In [897]:
#Outlier Detection (Boxplots)
plt.figure(figsize=(12, 4))
sns.boxplot(x=df["Amount"])
plt.title("Outliers in Transaction Amount")
# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\outliers_transactionamount.png')
plt.close()


In [898]:
#Investigating Extreme Values
df.sort_values("Amount").head(5)


TransactionId         BatchId       AccountId       SubscriptionId       CustomerId CurrencyCode CountryCode    ProviderId     ProductId ProductCategory    ChannelId     Amount    Value      TransactionStartTime  PricingStrategy  FraudResult  transaction_year  transaction_month  transaction_day  transaction_hour
55763   TransactionId_33290   BatchId_33774  AccountId_4249  SubscriptionId_4429  CustomerId_7338          UGX         256  ProviderId_4   ProductId_3         airtime  ChannelId_2 -1000000.0  1000000 2019-01-10 14:23:21+00:00                4            0              2019                  1               10                14
55757   TransactionId_64133  BatchId_100855  AccountId_4249  SubscriptionId_4429  CustomerId_7338          UGX         256  ProviderId_4  ProductId_10         airtime  ChannelId_2 -1000000.0  1000000 2019-01-10 14:17:57+00:00                4            0              2019                  1               10                14
55822   TransactionId_56954   BatchId_76447  AccountId_4249  SubscriptionId_4429  CustomerId_7339          UGX         256  ProviderId_4   ProductId_3         airtime  ChannelId_2  -900000.0   900000 2019-01-10 15:48:59+00:00                4            1              2019                  1               10                15
60834  TransactionId_126320   BatchId_93368  AccountId_4249  SubscriptionId_4429  CustomerId_7429          UGX         256  ProviderId_4   ProductId_3         airtime  ChannelId_2  -500000.0   500000 2019-01-15 11:09:57+00:00                0            1              2019                  1               15                11
55841   TransactionId_73797  BatchId_127642  AccountId_4249  SubscriptionId_4429  CustomerId_7339          UGX         256  ProviderId_4  ProductId_10         airtime  ChannelId_2  -500000.0   500000 2019-01-10 16:03:35+00:00                4            1              2019                  1               10                16

In [899]:
df["Amount_log"] = np.sign(df["Amount"]) * np.log1p(np.abs(df["Amount"]))


In [900]:
import matplotlib.pyplot as plt

plt.figure(figsize=(8, 4))
plt.hist(df['Amount'], bins=100)
plt.title('Transaction Amount Distribution')
plt.xlabel('Amount')
plt.ylabel('Frequency')
# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\Transaction_Amount_Distribution.png')
plt.close()


In [901]:
plt.figure(figsize=(8, 2))
plt.boxplot(df['Amount'], vert=False)
plt.title('Boxplot of Transaction Amount')
# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\Boxplot_Transaction_Amount.png')
plt.close()



In [902]:

# 4. SUMMARY STATISTICS


df.describe(include="all")


TransactionId        BatchId       AccountId       SubscriptionId       CustomerId CurrencyCode  CountryCode    ProviderId    ProductId     ProductCategory    ChannelId        Amount         Value                 TransactionStartTime  PricingStrategy   FraudResult  transaction_year  transaction_month  transaction_day  transaction_hour    Amount_log
count                 95662          95662           95662                95662            95662        95662      95662.0         95662        95662               95662        95662  9.566200e+04  9.566200e+04                                95662     95662.000000  95662.000000      95662.000000       95662.000000     95662.000000      95662.000000  95662.000000
unique                95662          94809            3633                 3627             3742            1          1.0             6           23                   9            4           NaN           NaN                                  NaN              NaN           NaN               NaN                NaN              NaN               NaN           NaN
top     TransactionId_76871  BatchId_67019  AccountId_4841  SubscriptionId_3829  CustomerId_7343          UGX        256.0  ProviderId_4  ProductId_6  financial_services  ChannelId_3           NaN           NaN                                  NaN              NaN           NaN               NaN                NaN              NaN               NaN           NaN
freq                      1             28           30893                32630             4091        95662      95662.0         38189        32635               45405        56935           NaN           NaN                                  NaN              NaN           NaN               NaN                NaN              NaN               NaN           NaN
mean                    NaN            NaN             NaN                  NaN              NaN          NaN          NaN           NaN          NaN                 NaN          NaN  6.717846e+03  9.900584e+03  2019-01-02 01:15:52.886433536+00:00         2.255974      0.002018       2018.497115           6.566233        15.902898         12.447722      2.502228
min                     NaN            NaN             NaN                  NaN              NaN          NaN          NaN           NaN          NaN                 NaN          NaN -1.000000e+06  2.000000e+00            2018-11-15 02:18:49+00:00         0.000000      0.000000       2018.000000           1.000000         1.000000          0.000000    -13.815512
25%                     NaN            NaN             NaN                  NaN              NaN          NaN          NaN           NaN          NaN                 NaN          NaN -5.000000e+01  2.750000e+02  2018-12-12 09:47:57.750000128+00:00         2.000000      0.000000       2018.000000           1.000000         8.000000          8.000000     -3.931826
50%                     NaN            NaN             NaN                  NaN              NaN          NaN          NaN           NaN          NaN                 NaN          NaN  1.000000e+03  1.000000e+03            2018-12-31 17:20:45+00:00         2.000000      0.000000       2018.000000          11.000000        16.000000         13.000000      6.908755
75%                     NaN            NaN             NaN                  NaN              NaN          NaN          NaN           NaN          NaN                 NaN          NaN  2.800000e+03  5.000000e+03     2019-01-25 10:54:47.500000+00:00         2.000000      0.000000       2019.000000          12.000000        24.000000         17.000000      7.937732
max                     NaN            NaN             NaN                  NaN              NaN          NaN          NaN           NaN          NaN                 NaN          NaN  9.880000e+06  9.880000e+06            2019-02-13 10:01:28+00:00         4.000000      1.000000       2019.000000          12.000000        31.000000         23.000000     16.106023

In [903]:

# 5. DISTRIBUTION OF NUMERIC FEATURES


numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()

df[numeric_cols].hist(figsize=(15, 12), bins=40)
plt.suptitle("Distribution of Numerical Features", fontsize=16)
# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\Distribution_Numerical_Features.png')
plt.close()


In [904]:

# 6. DISTRIBUTION OF CATEGORICAL FEATURES


categorical_cols = df.select_dtypes(include=["object"]).columns.tolist()

for col in categorical_cols:
    plt.figure(figsize=(10,4))
    df[col].value_counts().head(10).plot(kind='bar')
    plt.title(f"Top Categories in {col}")
    # Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\Top_Categories.png')
plt.close()


In [905]:

# 7. CORRELATION ANALYSIS


plt.figure(figsize=(12, 8))
corr = df[numeric_cols].corr()

sns.heatmap(corr, cmap="viridis", annot=False)
plt.title("Correlation Heatmap")
 # Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\Correlation_Heatmap.png')
plt.close()

corr


Amount     Value  PricingStrategy  FraudResult  transaction_year  transaction_month  transaction_day  transaction_hour  Amount_log
Amount             1.000000  0.989692        -0.061931     0.557370          0.010738          -0.009060        -0.003905         -0.001611    0.119257
Value              0.989692  1.000000        -0.017020     0.566739          0.012001          -0.010452        -0.024104          0.000474    0.073536
PricingStrategy   -0.061931 -0.017020         1.000000    -0.033821          0.029757          -0.031291        -0.131597          0.007423   -0.122450
FraudResult        0.557370  0.566739        -0.033821     1.000000          0.009811          -0.008887        -0.008636          0.008295    0.068706
transaction_year   0.010738  0.012001         0.029757     0.009811          1.000000          -0.996205        -0.247493         -0.009621    0.046641
transaction_month -0.009060 -0.010452        -0.031291    -0.008887         -0.996205           1.000000         0.207837          0.012241   -0.046175
transaction_day   -0.003905 -0.024104        -0.131597    -0.008636         -0.247493           0.207837         1.000000         -0.019464    0.057829
transaction_hour  -0.001611  0.000474         0.007423     0.008295         -0.009621           0.012241        -0.019464          1.000000   -0.008594
Amount_log         0.119257  0.073536        -0.122450     0.068706          0.046641          -0.046175         0.057829         -0.008594    1.000000

In [906]:

# 8. MISSING VALUES

missing = df.isnull().mean().sort_values(ascending=False)

plt.figure(figsize=(10,5))
missing.plot(kind="bar")
plt.title("Percentage of Missing Data by Feature")
plt.ylabel("Missing %")
# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\Percentage_MissingData_Feature.png')
plt.close()

missing


TransactionId           0.0
BatchId                 0.0
AccountId               0.0
SubscriptionId          0.0
CustomerId              0.0
CurrencyCode            0.0
CountryCode             0.0
ProviderId              0.0
ProductId               0.0
ProductCategory         0.0
ChannelId               0.0
Amount                  0.0
Value                   0.0
TransactionStartTime    0.0
PricingStrategy         0.0
FraudResult             0.0
transaction_year        0.0
transaction_month       0.0
transaction_day         0.0
transaction_hour        0.0
Amount_log              0.0
dtype: float64

In [907]:

# 9. OUTLIER DETECTION
# (using boxplots)


for col in numeric_cols:
    plt.figure(figsize=(8, 3))
    sns.boxplot(x=df[col], color="purple")
    plt.title(f"Outlier Check: {col}")


In [908]:
num_cols = df.select_dtypes(include="number")

plt.figure(figsize=(10, 6))
sns.heatmap(num_cols.corr(), cmap="coolwarm", center=0)
plt.title("Correlation Matrix")
# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\Correlation_Matrix.png')
plt.close()

In [909]:
import matplotlib.pyplot as plt
#Distribution of Numerical Features
num_cols = ["Amount", "Value"]

for col in num_cols:
    plt.figure(figsize=(6,4))
    sns.histplot(df[col], bins=100, kde=True)
    plt.title(f"Distribution of {col}")
# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\Distribution.png')
plt.close()


In [910]:
#RFM Feature Engineering (Core Risk Signal)
rfm = df.groupby("CustomerId").agg({
    "TransactionStartTime": lambda x: (df["TransactionStartTime"].max() - x.max()).days,
    "TransactionId": "count",
    "Amount": "sum"
}).reset_index()

rfm.columns = ["CustomerId", "Recency", "Frequency", "Monetary"]
rfm.head()


CustomerId  Recency  Frequency  Monetary
0     CustomerId_1       83          1  -10000.0
1    CustomerId_10       83          1  -10000.0
2  CustomerId_1001       89          5   20000.0
3  CustomerId_1002       25         11    4225.0
4  CustomerId_1003       11          6   20000.0

In [911]:
#RFM Correlation Analysis
sns.heatmap(rfm[["Recency","Frequency","Monetary"]].corr(), annot=True, cmap="coolwarm")
plt.title("RFM Correlation Heatmap")
# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\RFM_Correlation_Heatmap.png')
plt.close()


In [912]:
def test_rfm_shape():
    rfm = build_rfm(sample_df)
    assert all(col in rfm.columns for col in ["Recency", "Frequency", "Monetary"])


# Exploratory Data Analysis (EDA) - Key Insights

This section summarizes the key insights from the EDA conducted on the credit risk dataset.

---

## 1. Dataset Overview
- **Total transactions:** 95,662  
- **Currency:** All transactions in `UGX`  
- **Top Account IDs:**  
  - `AccountId_4841`: 30,893 transactions  
  - `AccountId_4249`: 4,457 transactions  
- **Time span:** 2018–2019, with transactions distributed across months, days, and hours

---

## 2. Numerical Features
- **Amount & Value:**  
  - Mean: ~6,718 and ~9,901 respectively  
  - High standard deviation (~123,000) with extreme negative and positive values  
  - Log-transformed `Amount_log` reduces skewness and normalizes the distribution
- **PricingStrategy:** Mostly value `2`, low variance  
- **FraudResult:** Very low fraud occurrence (~0.2%)

---

## 3. Transaction Temporal Patterns
- **Year:** Almost equal distribution between 2018 and 2019  
- **Month:** Higher transaction frequency in January, November, and December  
- **Day:** Fairly uniform distribution across days of the month  
- **Hour:** Peak hours between 8 AM–5 PM, some outliers at 0–3 and 22–23 hours  

---

## 4. Correlation Insights
- `Amount` and `Value` are **highly correlated (0.99)** → one can be used for modeling  
- `FraudResult` moderately correlated with `Amount` and `Value` (~0.56) → larger transactions slightly more likely to be fraudulent  
- Temporal features mostly uncorrelated with fraud

---

## 5. Outlier Detection
- Extreme values in `Amount` and `Amount_log` detected  
- `transaction_hour` mostly 8–17 hours, outliers at early morning (0–3) and late night (22–23)  
- Negative or extremely large amounts need attention before modeling

---

## 6. Categorical Insights
- Single `CountryCode` (`UGX`)  
- Top `AccountId`s dominate the dataset → handle high-cardinality carefully during encoding

---

## 7. Actionable Insights
1. Cap or transform outliers in `Amount` and `Value`  
2. Use `Amount_log` to reduce skewness  
3. Address low fraud ratio via resampling or weighting  
4. Transform temporal features into bins (morning, afternoon, evening)  
5. Consider dimensionality reduction due to high correlation between `Amount` and `Value`

---


# Feature Engineering
## targeted columns from the dataset for feature engineering 
- CustomerId               → customer identifier
- Amount                   → transaction amount
- TransactionStartTime     → transaction timestamp
- FraudResult              → binary target

In [913]:
#Feature Engineering
import pandas as pd
import numpy as np

from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer

from xverse.transformer import WOE



# Custom Transformer: Datetime Feature Extraction
## Features Created
- Transaction Hour
- Day
- Month
- Year

In [914]:
# Custom Transformer: Datetime Feature Extraction
class DateTimeFeatures:
    """
    Extracts hour, day, month, year
    """

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        X = X.copy()
        X["tx_hour"] = X["TransactionStartTime"].dt.hour
        X["tx_day"] = X["TransactionStartTime"].dt.day
        X["tx_month"] = X["TransactionStartTime"].dt.month
        X["tx_year"] = X["TransactionStartTime"].dt.year
        return X


# Custom Transformer: Customer-Level Aggregation
## Aggregate Features (Per CustomerId)
- Total Amount
- Average Amount
- Transaction Count
- Std Deviation of Amount

In [915]:
# Custom Transformer: Customer-Level Aggregation
## Aggregate Features (Per CustomerId)
class TransactionAggregator:
    """
    Aggregates transactions per customer
    """

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        agg = (
            X.groupby("CustomerId")
            .agg(
                total_amount=("Amount", "sum"),
                avg_amount=("Amount", "mean"),
                transaction_count=("TransactionId", "count"),
                amount_std=("Amount", "std")
            )
            .fillna(0)
            .reset_index()
        )
        return agg



In [916]:
numeric_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

categorical_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False))
])


In [917]:
# Feature Groups of Numeric values 
NUMERIC_FEATURES = [
    "total_amount",
    "avg_amount",
    "transaction_count",
    "amount_std"
]





In [918]:
# Feature Groups of Categorical Features
CATEGORICAL_FEATURES = [
    "CountryCode",
    "CountryCode",
    "ProviderId",
    "ProductCategory",
    "ChannelId",
    "PricingStrategy"
]


In [919]:
#Numeric Pipeline (Imputation + Standardization)

numeric_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])


In [920]:
#Categorical Pipeline (Imputation + One-Hot Encoding)
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder

categorical_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(
        handle_unknown="ignore",
        sparse_output=False  # ✅ correct for sklearn >= 1.2
    ))
])


In [921]:
#Column Transformer
preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_pipeline, NUMERIC_FEATURES),
        ("cat", categorical_pipeline, CATEGORICAL_FEATURES)
    ],
    remainder="drop"
)




In [922]:
#WoE Transformer (Credit-Risk Standard)
woe_encoder = WOEEncoder(
    cols=None,            # Apply WoE to all features
    handle_missing="value",
    handle_unknown="value"
)



In [923]:
#FULL FEATURE ENGINEERING PIPELINE (FIXED)
feature_pipeline = Pipeline(steps=[
    ("datetime", DateTimeFeatures()),
    ("aggregation", TransactionAggregator()),
    ("preprocessing", preprocessor),
    ("woe", woe_encoder)
])



In [924]:
#Feature Engineering Runner
def run_feature_engineering(df, target_col="is_high_risk"):
    """
    Full feature pipeline
    """

    if target_col not in df.columns:
        raise ValueError(f"Target column '{target_col}' not found.")

    y = df[target_col]

    df = DateTimeFeatures().fit_transform(df)
    df_agg = TransactionAggregator().fit_transform(df)

    X = preprocessor.fit_transform(df_agg)

    return X, y


In [925]:
#End-to-End Runner Function
def run_feature_engineering(df, target_col="default"):
    """
    Executes the full feature engineering pipeline.
    """

    df = df.copy()

    # Separate target safely
    if target_col not in df.columns:
        raise ValueError(f"Target column '{target_col}' not found.")

    y = df[target_col]
    X = df.drop(columns=[target_col])

    # Fit + transform
    X_transformed = feature_pipeline.fit_transform(X, y)

    # Convert to DataFrame
    feature_names = feature_pipeline.named_steps["preprocessing"].get_feature_names_out()
    X_final = pd.DataFrame(X_transformed, columns=feature_names)

    return X_final, y


In [926]:
#End-to-End Execution Function (With Error Handling)
def run_feature_engineering(df, target_col="FraudResult"):
    try:
        X = df.drop(columns=[target_col])
        y = df[target_col]

        X_transformed = feature_engineering_pipeline.fit_transform(X, y)

        feature_names = woe_transformer.get_feature_names_out()

        X_final = pd.DataFrame(X_transformed, columns=feature_names)

        return X_final, y

    except Exception as e:
        raise RuntimeError(f"Feature engineering pipeline failed: {e}")


In [927]:
df.to_csv(r"C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\data\process\transactions_cleaned.csv", index=False)


In [928]:
# Core
import pandas as pd
import numpy as np

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Preprocessing & ML
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans

# Settings
pd.set_option("display.max_columns", None)
sns.set(style="whitegrid")


In [929]:
#Convert Transaction Time to Datetime
df["TransactionStartTime"] = pd.to_datetime(
    df["TransactionStartTime"], utc=True
)

df["TransactionStartTime"].dtype

datetime64[ns, UTC]

In [930]:
#Define Snapshot Date
# Snapshot date = one day after the last transaction
snapshot_date = df["TransactionStartTime"].max() + pd.Timedelta(days=1)

snapshot_date

Timestamp('2019-02-14 10:01:28+0000', tz='UTC')

In [931]:
#Calculate RFM Metrics
rfm = (
    df
    .groupby("CustomerId")
    .agg(
        Recency=("TransactionStartTime",
                 lambda x: (snapshot_date - x.max()).days),
        Frequency=("TransactionId", "count"),
        Monetary=("Amount", "sum")
    )
    .reset_index()
)

rfm.head()


CustomerId  Recency  Frequency  Monetary
0     CustomerId_1       84          1  -10000.0
1    CustomerId_10       84          1  -10000.0
2  CustomerId_1001       90          5   20000.0
3  CustomerId_1002       26         11    4225.0
4  CustomerId_1003       12          6   20000.0

In [932]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler


def compute_rfm(df: pd.DataFrame, snapshot_date: pd.Timestamp) -> pd.DataFrame:
rfm = df.groupby("CustomerId").agg(
recency=("TransactionStartTime", lambda x: (snapshot_date - x.max()).days),
frequency=("TransactionId", "count"),
monetary=("Amount", "sum"),
).reset_index()

IndentationError: expected an indented block after function definition on line 5 (1112685731.py, line 6)

In [ ]:
#Inspect RFM Distributions
fig, axes = plt.subplots(1, 3, figsize=(18, 5))

sns.histplot(rfm["Recency"], bins=30, ax=axes[0])
axes[0].set_title("Recency Distribution")

sns.histplot(rfm["Frequency"], bins=30, ax=axes[1])
axes[1].set_title("Frequency Distribution")

sns.histplot(rfm["Monetary"], bins=30, ax=axes[2])
axes[2].set_title("Monetary Distribution")

plt.tight_layout()
# Save the plot as an image
plt.savefig(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\output\Proxy Target Variable Engineering\Inspect_RFM_Distributions.png')
plt.close()


In [ ]:
#Scale RFM Features
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

rfm_scaled = scaler.fit_transform(
    rfm[["Recency", "Frequency", "Monetary"]]
)


In [ ]:
#K-Means Clustering
from sklearn.cluster import KMeans

kmeans = KMeans(
    n_clusters=3,
    random_state=42,
    n_init=10
)

rfm["cluster"] = kmeans.fit_predict(rfm_scaled)
rfm.head()

CustomerId  Recency  Frequency  Monetary  cluster
0     CustomerId_1       84          1  -10000.0        0
1    CustomerId_10       84          1  -10000.0        0
2  CustomerId_1001       90          5   20000.0        0
3  CustomerId_1002       26         11    4225.0        1
4  CustomerId_1003       12          6   20000.0        1

In [ ]:
#Analyze Clusters
cluster_summary = (
    rfm
    .groupby("cluster")[["Recency", "Frequency", "Monetary"]]
    .mean()
    .sort_values("Recency", ascending=False)
)

cluster_summary


Recency    Frequency      Monetary
cluster                                      
0        61.877279     7.720196  8.172068e+04
2        29.000000  4091.000000 -1.049000e+08
1        12.726566    34.800000  2.725741e+05

In [ ]:
#Define Proxy Target Variable
# Identify high-risk cluster (worst engagement)
high_risk_cluster = cluster_summary.index[0]

rfm["is_high_risk"] = (rfm["cluster"] == high_risk_cluster).astype(int)

rfm[["CustomerId", "cluster", "is_high_risk"]].head()


CustomerId  cluster  is_high_risk
0     CustomerId_1        0             1
1    CustomerId_10        0             1
2  CustomerId_1001        0             1
3  CustomerId_1002        1             0
4  CustomerId_1003        1             0

In [ ]:
#Merge Target Back to Main Dataset
df_model = df.merge(
    rfm[["CustomerId", "is_high_risk"]],
    on="CustomerId",
    how="left"
)

df_model["is_high_risk"].value_counts(normalize=True)


is_high_risk
0    0.884918
1    0.115082
Name: proportion, dtype: float64

In [ ]:
#Save Final Modeling Dataset
df_model.to_csv(
    r"C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\data\process\transactions_with_risk_target.csv",
    index=False
)

print("✅ Proxy target variable successfully created and saved.")


✅ Proxy target variable successfully created and saved.


In [ ]:
import pandas as pd

# Example data
data = {
    'TransactionStartTime': ['2023-01-01', '2023-01-02'],
    'TransactionValue': [100, 150]
}
rfm_df = pd.DataFrame(data)

In [ ]:
rfm_df.to_csv(r'C:\Users\hp\Pictures\Bati Bank\Credit-Risk-Probability-Model-for-Alternative-Data\data\process\rfm_data.csv', index=False)

In [ ]:
#Define Reference Date
snapshot_date = rfm_df["TransactionStartTime"].max() + pd.Timedelta(days=1)
snapshot_date

TypeError: can only concatenate str (not "Timedelta") to str

In [ ]:
# Core
import numpy as np
import pandas as pd

# Modeling
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

# Metrics
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    roc_curve
)

# Experiment Tracking
import mlflow
import mlflow.sklearn

# Reproducibility
RANDOM_STATE = 42


In [ ]:
#Train–Test Split (Now Correct)
#We split after feature engineering and proxy target creation.
X_train, X_test, y_train, y_test = train_test_split(
    X_model,
    y_model,
    test_size=0.2,
    random_state=RANDOM_STATE,
    stratify=y_model
)

print("Train shape:", X_train.shape)
print("Test shape:", X_test.shape)


NameError: name 'X_model' is not defined

In [ ]:
#Train / Test Split
#We split data after feature engineering and proxy target creation
# X_model and y_model come from Task 3 + Task 4
X_train, X_test, y_train, y_test = train_test_split(
    X_model,
    y_model,
    test_size=0.2,
    random_state=42,
    stratify=y_model
)

print("Train shape:", X_train.shape)
print("Test shape:", X_test.shape)


NameError: name 'X_model' is not defined